### PR TITLE
SET = instead of + if value is null

### PR DESCRIPTION
**Pull request by bjalder26**
This addresses issue #727

It would make it so the JSON below would give you "property1": "new text" instead of "property1": "nullnew text" if the property was a null value previously.
{
"func": "SET",
"property": "property1",
"relation": "+",
"value": "new text",
"collection": "thisButton"
}

Updated to NOT perform a SET if the relationship is `+` and the new value is `null` to prevent results where the property gets set to "old valuenull" instead of remaining "old value".

Game-breaking potential:

- `SET` `+` is mistakenly used on a `null` property instead of `SET` `=`
- The property is then used as a operand in an `IF` statement
- The 'working' behavior is for the property to NOT be equal to what it was just `SET` to be
OR
NOTE: with conditions, currently `SET` `+` `false` to a `null` value gives 0 and   `SET` `+` `true` to a `null` value gives 1, so they should operate the same.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3728/pr-test (or any other room on that server)

**Comment by rogerl50** [comment]
Why is this specific to `+`? How about the other relations? Are there any cases where we wouldn't want to simply convert to `=`?

**Comment by bjalder26** [comment]
> Why is this specific to `+`? How about the other relations? Are there any cases where we wouldn't want to simply convert to `=`?

The reason why with + is because it was concatenating new values with `null`.  So if there happened to be no value, and you use `+` to add 'this text' not being able to know beforehand that it would be null, then you would get 'nullthis text'.  In other words, because as far as concatenating text goes, when no value is shown it actually has a "hidden" text value of "null".

My thought was that if a game designer added (`+`) text to what appears to be nothing (null), they would expect to just get the same value back, and certainly not the value appended to `null`.

I can't think of another relation that would work that way, but I could be missing something.

**Comment by rogerl50** [comment]
> I can't think of another relation that would work that way, but I could be missing something.

Fair enough; I tested other relations and they work in a reasonable way.

How about the (admittedly unlikely) case where the property being modified has a value, but the value being added is null? At present, this produces `'oldvaluenull'`. Here, just setting the relation to `=` probably gives unexpected results; perhaps setting the second value to `''` is the best option? I'm not sure. Or perhaps we should just leave it alone.

**Comment by bjalder26** [comment]
> > I can't think of another relation that would work that way, but I could be missing something.
> 
> Fair enough; I tested other relations and they work in a reasonable way.
> 
> How about the (admittedly unlikely) case where the property being modified has a value, but the value being added is null? At present, this produces `'oldvaluenull'`. Here, just setting the relation to `=` probably gives unexpected results; perhaps setting the second value to `''` is the best option? I'm not sure. Or perhaps we should just leave it alone.

Good catch.  I'm thinking that in that case there should not be a SET at all.  I think that would give the expected result in all cases.

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -1065,7 +1065,8 @@ export class Widget extends StateManaged {
           for(const w of collections[a.collection]) {
             if(a.relation == '+' && w.get(String(a.property)) == null)
               a.relation = '=';
-            await w.set(String(a.property), compute(a.relation, null, w.get(String(a.property)), a.value));
+             if(!(a.relation == '+' && a.value == null))
```
Comment (rogerl50) [inline client/js/widgets/widget.js:1068]: 1) Indentation is off on the newly added if (and the subsequent statement).
2) It is probably worthwhile to add something like
```
            problems.push(`null value being appended, SET ignored`)
```

**Review by rogerl50 — CHANGES_REQUESTED** [review]

**Comment by rogerl50** [comment]
BTW this is still marked Draft...

**Comment by bjalder26** [comment]
> BTW this is still marked Draft...

I made it draft again when I realized I should make the change you pointed out.  Now I somehow lost my test room :).

**Pull request merged**